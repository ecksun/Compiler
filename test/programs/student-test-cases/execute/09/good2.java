
/* Creates an array and quicksorts it. Extensions:
   /*      variable declarations allowed in main (Code:VDM)
   /*      operators >, >=, <=, ||, ==, !=  
   /*      inheritance
   /*      for loops              (Code: FOR)
   /*      i++ and i-- statements (Code: PLM)
   /*      System.out.println() can take int[], boolean, and object type (Code: PRT)
*/
//EXT:ISC
//EXT:ICG
//EXT:JVM
//EXT:IWE
//EXT:CLE
//EXT:CGT
//EXT:CGE
//EXT:CEQ
//EXT:CNE
//EXT:BDJ
//EXT:PLM
//EXT:FOR
//EXT:VDM
//EXT:PRT

class Good2 {

    public static void main(String[] args){
         
        int[] a;
        int b;
        Good2Sorter s;
        int waste;
        
        a=new int[5];
        a[0]=7;
        a[1]=2;
        a[2]=1;
        a[3]=3;
        a[4]=10;
        
        s = new Good2Sorter();
        System.out.println(a);
        waste=s.quickSort(a,0,a.length-1);
        System.out.println(a);
        System.out.println(a[3]==a[4]);
        System.out.println(a[2]!=a[1]);
        
        for (b = a.length -1; b >= 0; b--)
            System.out.println((a[b]==3) || (b==2));
        if (a[a.length-1] > b)
            System.out.println(b);
    }
}

class Good2Bucket {
    int ref;
    int lb;
    int hb;
    int tmp;
    int dump;
}

class Good2Sorter extends Good2Bucket {

    public int quickSort(int[] array, int start, int end){
        int ret;
        ret = 0;
        if ( 1 > (end-start) ){
            ret = 0;
        } else {
            ref=array[start];
            lb=start;
            hb=end+1;
            
            while(lb+1 <= hb-1 ){
                if( (array[lb+1]-ref) < 0 ){
                    lb++;
                }
                else {
                    hb--;
                    tmp=array[hb];
                    array[hb]=array[lb+1];
                    array[lb+1]=tmp;
                }
            }
            array[start]=array[lb];
            array[lb]=ref;
            dump = this.quickSort(array,start,lb-1);
            dump = this.quickSort(array,lb+1,end);
            ret = 0;
        }
        return ret;
    }
}