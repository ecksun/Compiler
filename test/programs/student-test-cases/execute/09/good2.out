7 2 1 3 10 
1 2 3 7 10 
false
true
false
false
true
false
false
-1
